openedx/core/djangoapps/programs/utils.py

```diff
@@ -132,28 +133,120 @@ def get_display_category(program):
     return display_candidate
 
 
-def get_engaged_programs(user, enrollments):
-    """Derive a list of programs in which the given user is engaged.
+class ProgramProgressMeter(object):
+    """Utility for gauging a user's progress towards program completion.
 
     Arguments:
         user (User): The user for which to find programs.
-        enrollments (list): The user's enrollments.
-
-    Returns:
-        list of serialized programs, ordered by most recent enrollment
+        enrollments (list): The user's active enrollments.
     """
-    programs = get_programs(user)
-
-    enrollments = sorted(enrollments, key=lambda e: e.created, reverse=True)
-    # enrollment.course_id is really a course key.
-    course_ids = [unicode(e.course_id) for e in enrollments]
+    def __init__(self, user, enrollments):
+        self.user = user
+        self.enrollments = sorted(enrollments, key=lambda e: e.created, reverse=True)
+
+        # enrollment.course_id is really a course key ಠ_ಠ
+        self.course_ids = [unicode(e.course_id) for e in self.enrollments]
+        self.engaged_programs = None
+        self.course_certs = None
+
+    @property
+    def programs(self):
+        """Derive a list of programs in which the given user is engaged.
+
+        Returns:
+            list of program dicts, ordered by most recent enrollment.
+        """
+        programs = get_programs(self.user)
+        flattened = flatten_programs(programs, self.course_ids)
+
+        self.engaged_programs = []
+        for course_id in self.course_ids:
+            for program in flattened.get(course_id, []):
+                if program not in self.engaged_programs:
+                    self.engaged_programs.append(program)
+
+        return self.engaged_programs
+
+    @property
+    def progress(self):
+        """Gauge a user's progress towards program completion.
+
+        Returns:
+            list of dict, each containing information about a user's progress
+                towards completing a program.
+        """
+        if self.engaged_programs is None:
+            self.programs  # pylint: disable=pointless-statement
```

Comment:
I think it's better to avoid silly statements like this. What about create a program initialization method and call that in the constructor? Then the property can be already populated and used.